### PR TITLE
Reimplement watch service

### PR DESCRIPTION
**Pull request by arturbosch**
Provides three commands:
- `project [path]` - sets the current project to analyze
- `analyze [subPath]?` - plain detekt run, super fast after first use
- `watch [start?|stop]` - watches path set by project command and runs detekt on each changed file

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -46,6 +46,7 @@ allprojects {
 		mavenLocal()
 		jcenter()
 		maven(url = "http://dl.bintray.com/jetbrains/spek")
+		maven(url = "https://dl.bintray.com/arturbosch/generic")
```
Comment (schalkms) [inline build.gradle.kts]: new util package?

File: detekt-watcher/README.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+## detekt-watcher - continuous static analysis
+
+Provides three commands:
+- `project [path]` - sets the current project to analyze
+- `analyze [subPath]?` - plain detekt run, super fast after first use
+- `watch [start?|stop]` - watches path set by project command and runs detekt on each changed file
```
Comment (schalkms) [inline detekt-watcher/README.md]: good!

File: detekt-watcher/build.gradle.kts [hunk 3]
```diff
@@ -8,12 +10,18 @@ configurations.testImplementation.extendsFrom(configurations.kotlinTest)
 val jcommanderVersion: String by project
 val junitPlatformVersion: String by project
 val spekVersion: String by project
+val kshVersion: String by project
 
 dependencies {
 	implementation("com.beust:jcommander:$jcommanderVersion")
+	implementation("io.gitlab.arturbosch:ksh:$kshVersion")
```
Comment (schalkms) [inline detekt-watcher/build.gradle.kts]: nice+clean lib

File: detekt-watcher/src/main/kotlin/io/gitlab/arturbosch/detekt/watcher/commands/Watch.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+				val watchedDir = WatchedDir(watchKey.isValid, watchedPath, events)
+
+				detekt.check(watchedDir)
+				watchKey.reset()
+			}
+		}
+	}
+}
+
+private const val TIMEOUT = 5L
```
Comment (schalkms) [inline detekt-watcher/src/main/kotlin/io/gitlab/arturbosch/detekt/watcher/commands/Watch.kt]: Should we make this configurable? I think we discussed the effects of the `TIMEOUT` when you introduced this function to detekt.

File: detekt-watcher/src/main/kotlin/io/gitlab/arturbosch/detekt/watcher/config/DetektHome.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import io.gitlab.arturbosch.kutils.ApplicationHomeFolder
+import java.nio.file.Path
+
+/**
+ * @author Artur Bosch
+ */
+class DetektHome(home: Path) : ApplicationHomeFolder(home) {
+
+	internal val configFile: Path = resolveFile(CONFIG_FILE, shouldCreate = true)
+	internal val historyFile: String = System.getProperty(USER_HOME) + HISTORY_FILE
```
Comment (schalkms) [inline detekt-watcher/src/main/kotlin/io/gitlab/arturbosch/detekt/watcher/config/DetektHome.kt]: `historyFile = logFile` ?

File: build.gradle.kts (hunk 1 above)
Comment (arturbosch) [reply build.gradle.kts]: Could also be moved to the watcher module. yes, its for ksh lib.

File: detekt-watcher/README.md (hunk 2 above)
Comment (arturbosch) [reply detekt-watcher/README.md]: Note: need to add more info like the builtin `help` commands

Comment (arturbosch) [reply detekt-watcher/README.md]: Note: add sentence about the "global" detekt configuration yaml file which can be put to `~/.detekt`

File: detekt-watcher/build.gradle.kts (hunk 3 above)
Comment (arturbosch) [reply detekt-watcher/build.gradle.kts]: thanks :)

**Pull request merged**